WIP: use `HashIntoType` and custom hashing instead of `u64`

ctb opened the pull request:
Brief description:

* implement threaded consume that chunks sequences into some num chunks, each of which is processed individually, and then combined at the end;
* add criterion benchmarking;
* build new k-mer hash type `HashIntoType` and provide custom (no-op) hashing for `HashIntoType`;
* pre-allocate tables with default capacity (currently `100_000`);
* replace `String` with `&str` where possible, to avoid copying/cloning;


notes for later:

fix lib when doing `cargo bench`:
```
ln -s ~/miniforge3/envs/py311/lib/libpython3.11.dylib target/release/deps/
```

relevant pyo3 issue:

https://github.com/PyO3/pyo3/issues/4155

TODO notes:
- [ ] consumed calculation/validation?
- [ ] custom hashing fn for kmers?

ctb committed:
cleanup & opt

ctb committed:
add bounds

ctb committed:
finish implementing

ctb committed:
mostly works

ctb committed:
implement our own hasher

ctb committed:
remove hashdiv stuff

ctb committed:
cargo fmt

ctb committed:
benchmarks now kinda working

ctb committed:
support both identity hashing and initial capacity

ctb committed:
set const capacity

ctb committed:
switch to using tuple struct

ctb committed:
clippy

ctb committed:
combine HashIntoType with IdentityHash

ctb committed:
comment and simplify

ctb committed:
upd

ctb committed:
add some tests of parallel consume

ctb committed:
Style fixes by Ruff

Adamtaranto commented:
@ctb can you post that link about how the magical custom hashing in rust works?

Adamtaranto committed:
trigger tests

Adamtaranto committed:
Style fixes by Ruff

Adamtaranto committed:
rm orcid

Adamtaranto committed:
rm orcid again

Adamtaranto committed:
disable benchmarks

Adamtaranto commented:
Planning to work on this in the near future. 

- Do we want to use [Murmurhash](https://crates.io/crates/murmurhash3) instead of custom hash? 
- Target file for Bench "consume" is missing, maybe? Does this need to be added to the repo? Have disabled for now to prevent build errors.

Adamtaranto commented:
@copilot+claude-opus-4.6 review edits in this PR and make a detailed plan to complete the goals. Identify tests that are required to ensure accurate output and compare efficiency to existing implementation.

Copilot committed:
Fix build, consumed tracking, parallel_consume2, cleanup debug comments, add tests

Copilot committed:
Remove unnecessary state.finish() call in Hash impl

Adamtaranto committed:
Change Python version to 3.12 in CI workflow

Updated Python version in CI workflow from 3.x to 3.12.

Copilot committed:
Add E. coli genome benchmarks using criterion

- Add criterion 0.8 dev-dependency
- Enable [[bench]] for genome benchmark target
- Create benches/genome.rs: benchmarks for consume, parallel_consume,
  and parallel_consume2 against the full E. coli GCF_000005845.2 genome
- Auto-download genome from NCBI FTP on first run; skip all benchmarks
  gracefully when network is unavailable

Copilot committed:
Remove unused niffler import in genome benchmark

Adamtaranto committed:
Change miniforge variant to Miniforge3

Adamtaranto committed:
Specify Python interpreter version in CI workflow

Adamtaranto committed:
style: format Rust code
